api/src/main/java/net/md_5/bungee/api/event/CommandsDeclareEvent.java

```diff
@@ -0,0 +1,58 @@
+package net.md_5.bungee.api.event;
+
+import com.mojang.brigadier.tree.RootCommandNode;
+import lombok.AccessLevel;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.Setter;
+import lombok.ToString;
+import net.md_5.bungee.api.connection.Connection;
+
+/**
+ * Event called when a backend server (on 1.13+) sends the command structure
+ * to a player, but before BungeeCord adds the dummy command nodes of
+ * registered commands.
+ *
+ * BungeeCord will not overwrite the modification made by the listeners.
+ */
+@Data
+@ToString(callSuper = true)
+@EqualsAndHashCode(callSuper = true)
+public class CommandsDeclareEvent extends TargetedEvent
+{
+    /**
+     * Wether or not the command tree is modified by this event.
+     *
+     * If this value is set to true, BungeeCord will ensure that the
+     * modifications made in the command tree, will be sent to the player.
+     * If this is false, the modifications may not be taken into account.
+     *
+     * When calling {@link #getRoot()}, this value is automatically set
+     * to true.
+     */
+    @Setter(value = AccessLevel.NONE)
+    private boolean modified = false;
+
+    /**
+     * The root command node of the command structure that will be send to the
+     * player.
+     */
+    private final RootCommandNode root;
```

Comment:
Should be a LiteralCommandNode of a CommandSender

Reply:
The event does not handle one command, but the full command structure, that is a `RootCommandNode` instance, not `LiteralCommandNode`. (see `Commands#getRoot()`)
However, using CommandSender may be a good way to parametize the raw type for brigadier class 👍

Reply:
Dude, you're overcomplicating it. Make it handle 1 command, call multiple times. And I don't think we should expose the root command node

Reply:
1. Why not expose the root command node ?
2. Calling the event for each command will make a big performance issue. Imagine, on the server, you have a lot of command (lets say you have Essentials and WorldEdit installed, that would be 100-150 commands ?), the event would be called that much time, going through the event handler every time. Also, It won’t be possible to add a new command with this kind of event.

At least, if you really don’t want to expose the root command node (I still have to understand why), we can provide a `List<LiteralCommandNode>` in the event, instead

Reply:
Alright then, keep it like that.